### PR TITLE
MegaGAM Checkout Draft

### PR DESCRIPTION
**Pull request by billy-yuan**
**What I did**
- Added Modal to the GAM page (also put a dummy button for testing that will be removed once we merge)
- Updated `ModalAmount` and `ModalCardDetails` to show different copies for Mega GAM `purchaseType`

**Outstanding**
- Need to add donation bar
- What seller info and other metadata needs to be passed through the checkout flow? Currently I am not passing any props or data through the checkout flow.
- Need to get updated copies
- How to check that payment flow works and the correct data is being sent?
- Update confirmation page

image
image
image

### CODE REVIEW

File: src/components/MerchantsPage/gam/GiftAMealPage.tsx [hunk 1]
```diff
@@ -7,14 +7,30 @@ import NoActiveCampaignsBox from './NoCampaignsBox';
...
+import Modal from '../../ModalPayment';
+import {
+  ModalPaymentConstants,
+  useModalPaymentDispatch,
+  ModalPaymentTypes,
+} from '../../../utilities/hooks/ModalPaymentContext';
+
+// todo: figure out what props are needed for MegaGAMModal
+// Props are for Modal are
+// sellerId: string;
```
Comment (billy-yuan) [inline src/components/MerchantsPage/gam/GiftAMealPage.tsx:19]: Forgot to mention donation amounts for the bar

File: src/consts.ts [hunk 2]
```diff
@@ -91,6 +91,7 @@ function hasKey<O>(obj: O, key: keyof any): key is keyof O {
 
 const LIGHT_UP_CHINATOWN_TIER_2_MIN = 45;
 const LIGHT_UP_CHINATOWN_TIER_3_MIN = 150;
+const MEGA_GAM = 'mega_gam';
```
Comment (aliceluo2048) [inline src/consts.ts:94]: What is the purpose of creating a constant? Could we use the enum instead?

Comment (billy-yuan) [reply src/consts.ts:94]: ah that's right. will change

File: src/components/MerchantsPage/gam/GiftAMealPage.tsx [hunk 3]
```diff
@@ -63,6 +87,9 @@ const GiftAMealPage = (props: Props) => {
       >
         {t('gamHome.backButton')}
       </button>
+      <MegaGAMModal></MegaGAMModal>
+      {/* Delete button below after testing */}
+      <button onClick={showModal}>GAM DONATE</button>
```
Comment (chrisng93) [inline src/components/MerchantsPage/gam/GiftAMealPage.tsx:92]: you're not merging this code in right?

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx [hunk 4]
```diff
@@ -66,28 +65,36 @@ export const Modal = (props: Props) => {
         return t('purchase.voucher', { seller: sellerName });
       case ModalPaymentTypes.modalPages.light_up_chinatown:
         return t('purchase.donation_to', { seller: sellerName });
+      case ModalPaymentTypes.modalPages.mega_gam:
+        return t(
+          'purchase.mega_gam',
+          sellerName ? { seller: sellerName } : { seller: 'Seller is blank' }
```
Comment (chrisng93) [inline src/components/ModalPayment/ModalAmount/ModalAmount.tsx:71]: we aren't going to have a seller for mega gam. i'm not sure where we're holding the "December Campaign" text. @justintmckibben we should probably add a name field to the campaign model for this OR we could use the project name

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx [hunk 5]
```diff
@@ -66,28 +65,36 @@ export const Modal = (props: Props) => {
...
+        ); // todo: confirm which copy to use & who the seller is
       default:
         return t('purchase.donation', { seller: sellerName });
     }
   };
 
+  const getSubheaderText = (purchaseType) => {
```
Comment (chrisng93) [inline src/components/ModalPayment/ModalAmount/ModalAmount.tsx:78]: is there a typing for this?

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx [hunk 6]
```diff
@@ -66,28 +65,36 @@ export const Modal = (props: Props) => {
...
-          `paymentProcessing.amount.${
-            modalView === ModalPaymentTypes.modalPages.light_up_chinatown
-              ? 'light_up_chinatown'
-              : 'header'
-          }`
-        )}
-      </p>
-
+      <SubHeader>{getSubheaderText(modalView)}</SubHeader>
+      <MegaGAMBar modalView={modalView}>MEGA GAM BAR</MegaGAMBar>
```
Comment (chrisng93) [inline src/components/ModalPayment/ModalAmount/ModalAmount.tsx:97]: are you going to delete this code before merging?

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx [hunk 7]
```diff
@@ -66,28 +65,36 @@ export const Modal = (props: Props) => {
...
+      default:
+        return t('paymentProcessing.amount.header');
+    }
+  };
+
   return (
     <ContentContainer id="donation-form" data-testid="modal-amount">
       <Header>{getHeaderText(modalView, props.sellerName)}</Header>
-
-      {props.sellerId === 'send-chinatown-love' && (
```
Comment (chrisng93) [inline src/components/ModalPayment/ModalAmount/ModalAmount.tsx:78]: why are we deleting this `props.sellerId === 'send-chinatown-love' ` code?

File: src/components/ModalPayment/ModalCardDetails/ModalCardDetails.tsx [hunk 8]
```diff
@@ -259,17 +268,16 @@ const ModalCardDetails = ({
...
-
-    if (
-      type === ModalPaymentTypes.modalPages.gift_card ||
-      (type === ModalPaymentTypes.modalPages.light_up_chinatown &&
-        amount >= LIGHT_UP_CHINATOWN_TIER_2_MIN)
-    ) {
-      return t('modalPayment.modalCardDetails.details.voucher');
-    } else {
-      return t('modalPayment.modalCardDetails.details.donation');
+    switch (true) {
```
Comment (chrisng93) [inline src/components/ModalPayment/ModalCardDetails/ModalCardDetails.tsx:271]: probably worth keeping the if statement syntax if we're not actually switching on anything here

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx [hunk 9]
```diff
@@ -66,28 +65,36 @@ export const Modal = (props: Props) => {
...
+          'purchase.mega_gam',
+          sellerName ? { seller: sellerName } : { seller: 'Seller is blank' }
+        ); // todo: confirm which copy to use & who the seller is
       default:
         return t('purchase.donation', { seller: sellerName });
     }
   };
 
+  const getSubheaderText = (purchaseType) => {
+    switch (purchaseType) {
```
Comment (chrisng93) [inline src/components/ModalPayment/ModalAmount/ModalAmount.tsx:79]: style nit:
```
switch (purchaseType) {
      case ModalPaymentTypes.modalPages.light_up_chinatown:
        return t('paymentProcessing.amount.light_up_chinatown');
      case ModalPaymentTypes.modalPages.mega_gam:
        return t('paymentProcessing.amount.mega_gam');
      case ModalPaymentTypes.modalPages.donation:
      case ModalPaymentTypes.modalPages.gift_card:
      default:
        return t('paymentProcessing.amount.header');
    }
}
```

File: src/locales/en/translation.json [hunk 10]
```diff
@@ -3,6 +3,7 @@
     "amount": {
```
Comment (chrisng93) [inline src/locales/en/translation.json:3]: you'll need to get the chinese translations as well

File: src/components/ModalPayment/ModalCardDetails/ModalCardDetails.tsx (hunk 8 above)
Comment (aliceluo2048) [reply src/components/ModalPayment/ModalCardDetails/ModalCardDetails.tsx:271]: unrelated but I was surprised we could use bool in a switch case

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx (hunk 6 above)
Comment (billy-yuan) [reply src/components/ModalPayment/ModalAmount/ModalAmount.tsx:97]: yup

File: src/components/ModalPayment/ModalCardDetails/ModalCardDetails.tsx (hunk 8 above)
Comment (billy-yuan) [reply src/components/ModalPayment/ModalCardDetails/ModalCardDetails.tsx:271]: no problem.i can convert it to if statements. I was trying to simplify the logic on the hand side but can easily switch the case statement to if statements

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx (hunk 7 above)
Comment (billy-yuan) [reply src/components/ModalPayment/ModalAmount/ModalAmount.tsx:78]: ` {props.sellerId === 'send-chinatown-love' && (
        <p>{t('donationPool.description2')}</p>
      )}`

So `donationPool.description2` is equal to `All donations will be distributed evenly to our merchants.",` but this is not used anywhere in the checkout modal. In addition, it isn't evaluating to a value that's being displayed on the page (unless I misunderstood this line!)

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx (hunk 9 above)
Comment (billy-yuan) [reply src/components/ModalPayment/ModalAmount/ModalAmount.tsx:79]: thanks, this is definitely cleaner

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx (hunk 5 above)
Comment (billy-yuan) [reply src/components/ModalPayment/ModalAmount/ModalAmount.tsx:78]: yep added

File: src/components/ModalPayment/ModalCardDetails/ModalCardDetails.tsx [hunk 11]
```diff
@@ -337,7 +341,11 @@ const ModalCardDetails = ({
               <b>
                 ${amount} {numberOfMealsText}
               </b>{' '}
-              to {sellerName}{' '}
+              {/* For megagam, confirm which sellerName to use in line below */}
+              to{' '}
+              {purchaseType === ModalPaymentTypes.modalPages.mega_gam
+                ? 'MEGA GAM - GET SELLER(S) INFO'
```
Comment (billy-yuan) [inline src/components/ModalPayment/ModalCardDetails/ModalCardDetails.tsx:347]: replace `MEGA GAM` with whatever the campaign name will be

File: src/components/ModalPayment/ModalAmount/ModalAmount.tsx (hunk 4 above)
Comment (billy-yuan) [reply src/components/ModalPayment/ModalAmount/ModalAmount.tsx]: got it. can wait until we decide that and then add it in

**Pull request closed without merging**